wingtips-core/src/main/java/com/nike/wingtips/Tracer.java

```diff
@@ -517,6 +543,98 @@ public void completeSubSpan() {
+     *                 managing spans we must assume the call was intentional. No error will be logged, and the span
+     *                 will be completed and logged to the "valid span logger".
+     *             </li>
+     *             <li>
+     *                 In either case, the current span stack and MDC info will be left untouched if the given span
+     *                 is not the current span.
+     *             </li>
+     *         </ul>
+     *     </li>
+     * </ul>
+     *
+     * <p>NOTE: This is intentionally package-scoped. Only {@link Span#close()} should ever call this method.
+     */
+    void handleSpanCloseMethod(Span span) {
+        // See if this span has already been completed - if so then this method should not have been called.
+        if (span.isCompleted()) {
+            classLogger.error(
+                "WINGTIPS USAGE ERROR - An attempt was made to close() a span that was already completed. "
+                + "This call to Span.close() will be ignored. "
+                + "wingtips_usage_error=true, already_completed_span=true, trace_id={}, span_id={}",
+                span.getTraceId(), span.getSpanId(), new Exception("Stack trace for debugging purposes")
+            );
+            return;
+        }
+
+        // See if this span is the current span.
+        if (span.equals(getCurrentSpan())) {
+            // This is the current span. Go ahead and complete it.
+            if (getCurrentSpanStackSize() > 1) {
+                // It's a subspan.
+                completeSubSpan();
+            }
+            else {
+                // It's an overall request span.
+                completeRequestSpan();
+            }
+        }
+        else {
+            // This is not the current span - something *might* be wrong with wingtips usage, or it could be a
+            //      span managed manually outside Tracer's view.
+            Deque<Span> currentSpanStack = currentSpanStackThreadLocal.get();
+            if (currentSpanStack != null && currentSpanStack.contains(span)) {
+                // This span is one being managed by Tracer but it's not the current one, therefore this is an invalid
+                //      wingtips usage situation.
+                classLogger.error(
+                    "WINGTIPS USAGE ERROR - An attempt was made to close() a Tracer-managed span that was not the "
+                    + "current span. This span will be completed as an invalid span but Tracer's current span stack "
+                    + "and the current MDC info will be left alone. "
+                    + "wingtips_usage_error=true, closed_non_current_span=true, trace_id={}, span_id={}",
+                    span.getTraceId(), span.getSpanId(), new Exception("Stack trace for debugging purposes")
+                );
+                completeAndLogSpan(span, true);
+            }
+            else {
```

Comment:
Ok so this is trying to figure out if a user called close twice (ex manually etc) this is the sort of code to watch out for

Reply:
If a span was closed twice the second one would be caught by the `if (span.isCompleted())` branch at the beginning of this method. This case here is when someone calls `span.close()` (or try-with-resources) on a span that is not in `Tracer`'s span stack on this thread - i.e. it's not "managed" by Tracer. It's a loose span. 

There are a few scenarios that could cause this, and some are error cases and others are not. 

* It could have been created manually using `Span`'s builder, intentionally bypassing `Tracer` because the user doesn't want to have to deal with the thread-local nature of `Tracer` and wants to handle `Span`s up front rather than behind-the-scenes. Useful for powerusers in certain highly asynchronous situations.
* Or it might have been created on a different thread with `Tracer` and passed around and not properly re-attached to `Tracer` on the new thread. In this case the user intended to use `Tracer` but failed to reattach, therefore it's an error case.

Since we can't know for sure in the code here whether it's a valid use case or not, we assume valid but spit out some debug logging so if a user is using it incorrectly they at least have a chance of being warned about it. 

It's a bit messy, but ... 🤷‍♂️